packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/methods/SuspendMethod.kt

```diff
@@ -0,0 +1,34 @@
+package expo.modules.kotlin.methods
+
+import expo.modules.kotlin.Promise
+import expo.modules.kotlin.exception.CodedException
+import expo.modules.kotlin.exception.UnexpectedException
+import expo.modules.kotlin.types.AnyType
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+
+class SuspendMethod(
+  name: String,
+  argsType: Array<AnyType>,
+  scopeProvider: Lazy<CoroutineScope>,
+  private val body: suspend CoroutineScope.(args: Array<out Any?>) -> Any?
+) : AnyMethod(name, argsType) {
+  private val scope by scopeProvider
+
+  // TODO(@lukmccall): improve error messages
+  override fun callImplementation(args: Array<out Any?>, promise: Promise) {
+    scope.launch {
+      try {
+        val result = body.invoke(scope, args)
+        if (isActive) {
+          promise.resolve(result)
+        }
+      } catch (e: CodedException) {
+        promise.reject(e)
+      } catch (e: Throwable) {
+        promise.reject(UnexpectedException(e))
+      }
+    }
```

Comment:
This try-catch won't handle uncaught exceptions in nested coroutines:
https://github.com/expo/expo/pull/16413#discussion_r817910242

See https://medium.com/androiddevelopers/exceptions-in-coroutines-ce8da1ec060c or the video embedded in there.

That's why I'd recommend `CoroutineExceptionHandler` here. You can write another test case to test this (see my comment in tests)

Reply:
Ok, as we talked priv, I might be wrong here so let's leave this as-is for now, I'll do more experiments on that when it lands